lib/min_heap.rb

```diff
@@ -44,28 +72,68 @@ def to_s
   end
 
   # This method returns true if the heap is empty
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(1) -- I *think* this code will stop running once 
+  # count gets higher than 0 but I'm not certain
+  # Space complexity: O(1)
   def empty?
-    raise NotImplementedError, "Method not implemented yet..."
+    @store.length == 0 ? true : false
   end
 
   private
 
   # This helper method takes an index and
   #  moves it up the heap, if it is less than it's parent node.
   #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(log n) because we're swapping max once per tree level
+  # Space complexity: O(log n)
   def heap_up(index)
+
+    # formula for determining parent index of a given element's index -- 
+    # parentheses are crucial!
+    parent_index = (index - 1) / 2
     
+    # BASE CASES
+    # stop if the element we're looking at is at the root of the tree
+    return if index == 0
+    # since this is a min heap, we're done if the parent's key is less than (or  
+    # equal to) the key of the element we're looking at
+    return if @store[parent_index].key <= @store[index].key
+
+    # RECURSIVE CASE: swap element with parent ...
+    @store[parent_index], @store[index] = @store[index], @store[parent_index]
+    # ... and make parent the new element we're heaping up
+    return heap_up(parent_index)
   end
 
   # This helper method takes an index and 
-  #  moves it up the heap if it's smaller
-  #  than it's parent node.
+  #  moves it down the heap if it's larger
+  #  than its child nodes.
   def heap_down(index)
-    raise NotImplementedError, "Method not implemented yet..."
+
+    left_child_index = 2 * index + 1
+    right_child_index = 2 * index + 2
+
+    # BASE CASES
+    # stop if we're at the last leaf in the tree 
+    return if @store[left_child_index].nil? && @store[right_child_index].nil?
+    # stop if both children are greater than or equal to parent
+    return if @store[left_child_index].key >= @store[index].key && @store
+
+    # RECURSIVE CASES
+    # if parent is greater than left child, swap and look at next level --
+    # need @store[index] conditional to stop code running after we've reached 
+    # the bottom of the heap
+    if @store[index] && @store[index].key > @store[left_child_index].key
```

Comment:
You also need to figure out which is smaller the left child or right child.  Otherwise you'll do some unnecessary heap_downs.